PyRoute/TradeCodes.py

```diff
@@ -26,6 +26,53 @@ class TradeCodes(object):
     pcolor = {'As': '#8E9397', 'De': '#d17533', 'Fl': '#e37dff', 'He': '#ff6f0c', 'Ic': '#A5F2F3',
               'Oc': '#0094ED', 'Po': '#6a986a', 'Va': '#c9c9c9', 'Wa': '#4abef4'}
     ext_codes = {'Lt', 'Ht', 'Lg', 'Hg'}
+    weird_codes = {'{Anomaly}', '{Fuel}', '{Ringworld}', '{Rosette}'}
+    allowed_residual_codes = {'Ag', 'Ba', 'Bo', 'C0', 'C1', 'C2', 'C3', 'C4', 'C5', 'C6', 'C7', 'C8', 'C9', 'Cl', 'Cw',
+                              'Cy', 'D0', 'D1', 'D2', 'D3', 'D4', 'D5', 'D6', 'D7', 'D8', 'D9', 'Dw', 'Ex', 'Fr', 'Ga',
+                              'Hi', 'In', 'Lo', 'N1', 'Na', 'Ni', 'o', 'Pa', 'Ph', 'Pi', 'Po', 'Pr', 'Ri', 'Rn', 'Rv',
+                              's', 'Sp', 'St', 'Tn', 'Za'}
+    ok_pairs = {('Ni', 'Po'), ('Ni', 'Pa'), ('Ni', 'Wa'), ('Ph', 'Ri'), ('Lo', 'Va'), ('Ni', 'Va'), ('Lo', 'Po'),
+                ('De', 'He'), ('De', 'Ni'), ('De', 'Po'), ('He', 'Ni'), ('He', 'Po'), ('Hi', 'In'), ('De', 'Lo'),
+                ('Ic', 'Na'), ('Ic', 'Ni'), ('Ic', 'Va'), ('Na', 'Ni'), ('Na', 'Va'), ('As', 'Ni'), ('As', 'Va'),
+                ('De', 'Na'), ('He', 'Na'), ('Na', 'Po'), ('Hi', 'Pr'), ('Ga', 'Hi'), ('Ag', 'Ni'), ('Na', 'Pi'),
+                ('Ni', 'Oc'), ('Ag', 'Ga'), ("Ag", "Ri"), ("Ga", "Ni"), ("Ga", "Ri"), ("Ni", "Ri"), ("Ag", "Pi"),
+                ("He", "Lo"), ("Ag", "Pr"), ("Ni", "Pr"), ("As", "Hi"), ("As", "In"), ("As", "Na"), ("Hi", "Na"),
+                ("Hi", "Va"), ("In", "Na"), ("In", "Va"), ("De", "Ph"), ("De", "Pi"), ("He", "Ph"), ("He", "Pi"),
+                ("Na", "Ph"), ("Ph", "Pi"), ("Ph", "Po"), ("Pi", "Po"), ("Ic", "Lo"), ("Fl", "He"), ("Fl", "Ni"),
+                ("As", "Lo"), ("Pa", "Ph"), ("Pa", "Ri"), ("Pa", "Pi"), ("Hi", "Wa"), ("In", "Wa"), ("Ga", "Lo"),
+                ("Ga", "Pa"), ("Ph", "Va"), ("Pi", "Va"), ("As", "Ph"), ("As", "Pi"), ("Ga", "Pr"), ("Ic", "Pi"),
+                ("Ba", "Va"), ("Hi", "Po"), ("Hi", "Ic"), ("Ic", "In"), ("Fl", "Lo"), ("De", "Ri"), ("Pi", "Wa"),
+                ("Ga", "Ph"), ("Ba", "Fl"), ("Ba", "Lo"), ("Ba", "Ni"), ("De", "Hi"), ("De", "In"), ("He", "Hi"),
```

Comment:
Ba Lo and Ba Ni are not allowed codes. It used to be the population codes would overlap and you could have "Ba Lo Ni" in your trade codes.

Reply:
Summarising existing practice, as above.  If a given code pair turned up at least once in the TravMap data, it got whitelisted here.